### PR TITLE
EntryGeneration: Classes and Functions

### PR DESCRIPTION
**Pull request by chippmann**
This is a first PR for the entry generation. It can register classes and functions.

It does not create any GDNS file yet.

### CODE REVIEW

File: entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/RpcModeAnnotationMapper.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+object RpcModeAnnotationMapper {
+    /**
+     * maps RpcModeAnnotations and RPCMode registration enums to ClassNames to provide import informations for kotlinPoet
+     */
+    fun mapRpcModeAnnotationToClassName(rpcModeAnnotationAsString: String): ClassName {
+        return when (rpcModeAnnotationAsString) {
+            "godot.registration.RPCMode.DISABLED" -> ClassName("godot.registration.RPCMode", "DISABLED")
+            "godot.registration.RPCMode.REMOTE" -> ClassName("godot.registration.RPCMode", "REMOTE")
+            "godot.registration.RPCMode.SYNC" -> ClassName("godot.registration.RPCMode", "SYNC")
```
Comment (piiertho) [inline entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/RpcModeAnnotationMapper.kt]: SYNC is deprecated.
I think deprecated should stay as they are not removed yet. But in this case, we miss SLAVE.
https://docs.godotengine.org/en/3.1/classes/class_multiplayerapi.html#enumerations

File: entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/transformer/TypeDeclarationsToClassWithMemberTransformer.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package godot.entrygenerator.transformer
+
+import godot.entrygenerator.model.ClassWithMembers
+import org.jetbrains.kotlin.descriptors.ClassDescriptor
+import org.jetbrains.kotlin.descriptors.FunctionDescriptor
+import org.jetbrains.kotlin.descriptors.PropertyDescriptor
+
+object TypeDeclarationsToClassWithMemberTransformer {
+
+    fun transform(
```
Comment (piiertho) [inline entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/transformer/TypeDeclarationsToClassWithMemberTransformer.kt]: I would rather prefer a top level method, but that's a personal preference.

File: godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/ClassHandle.kt [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    fun registerFunction(methodName: String, methodRef: COpaquePointer, rpcMode: RPCMode) {
+        disposables.add(methodRef)
+        memScoped {
+            val attribs = cValue<godot_method_attributes> {
+                rpc_type = when (rpcMode) {
+                    RPCMode.DISABLED -> GODOT_METHOD_RPC_MODE_DISABLED
+                    RPCMode.REMOTE -> GODOT_METHOD_RPC_MODE_REMOTE
+                    RPCMode.SYNC -> GODOT_METHOD_RPC_MODE_SYNC
+                    RPCMode.MASTER -> GODOT_METHOD_RPC_MODE_MASTER
+                    RPCMode.SLAVE -> GODOT_METHOD_RPC_MODE_SLAVE
```
Comment (piiertho) [inline godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/ClassHandle.kt]: Same here for RPC.

File: entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/RpcModeAnnotationMapper.kt (hunk 1 above)
Comment (chippmann) [reply entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/RpcModeAnnotationMapper.kt]: SLAVE was renamed to PUPPET for godot 3.1 i think

Comment (chippmann) [reply entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/RpcModeAnnotationMapper.kt]: and i think we can remove the deprecated ones

Comment (chippmann) [reply entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/RpcModeAnnotationMapper.kt]: i removed the deprecated one if it's ok for you

File: plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/GodotPlugin.kt [hunk 4]
```diff
@@ -24,7 +26,12 @@ class GodotPlugin : Plugin<Project> {
 
     private fun setupExtensionDefaults(project: Project, godot: GodotExtension) {
         with(godot) {
-            platforms(*Platform.values())
+            // we don't have godot-library in the mobile targets yet, limit these to desktop for now
```
Comment (piiertho) [inline plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/GodotPlugin.kt:29]: We should add a comment in both build file of godot-library and here to remember to add mobile targets in both, I think.

File: entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/transformer/TypeDeclarationsToClassWithMemberTransformer.kt (hunk 2 above)
Comment (chippmann) [reply entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/transformer/TypeDeclarationsToClassWithMemberTransformer.kt]: changed it

File: plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/GodotPlugin.kt (hunk 4 above)
Comment (chippmann) [reply plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/GodotPlugin.kt:29]: added

**Review by piiertho — CHANGES_REQUESTED** [review]
Few comment otherwise I'm ok. Really smart :)

**Review by piiertho — APPROVED** [review]
Ok for merging it ! :)

**Review by raniejade — APPROVED** [review]

**Pull request merged**